### PR TITLE
Usernames' support

### PR DESCRIPTION
**Pull request by MiroMargineanu**
### Issue/Feature
Add username support for the sign transaction screens.

### Reproduce
Issue exists on version `2.` of sdk-dapp.

### Contains breaking changes
[x] No
[] Yes

### Updated CHANGELOG
[x] Yes
[] No

### Testing
[x] User testing
[] Unit tests

### CODE REVIEW

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx [hunk 1]
```diff
@@ -1,34 +1,100 @@
...
   scamReport: string | null;
 }
 
-export const ConfirmReceiver = (props: ConfirmReceiverPropsType) => {
-  const { receiver, scamReport } = props;
+export const ConfirmReceiver = ({
+  receiver,
+  scamReport
+}: ConfirmReceiverPropsType) => {
+  const isSmartContract = isContract(receiver ?? '');
```
Comment (razvantomegea) [inline src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx:27]: Does it make sense to proceed with the code if receiver is missing?

File: src/utils/transactions/getUsernameForTransaction.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { isStringBase64 } from '../decoders';
+
+export const getUsernameForTransaction = (username?: string) => {
+  if (!username) {
+    return undefined;
```
Comment (razvantomegea) [inline src/utils/transactions/getUsernameForTransaction.ts:5]: return undefined looks bad. return; or return null;

**Review by razvantomegea — CHANGES_REQUESTED** [review]

File: src/utils/transactions/getUsernameForTransaction.ts (hunk 2 above)
Comment (MiroMargineanu) [reply src/utils/transactions/getUsernameForTransaction.ts]: Updated. 👍

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx (hunk 1 above)
Comment (MiroMargineanu) [reply src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx]: Updated. 👍

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,10 +6,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+- [Added usernames' support for the signing transaction screens](https://github.com/multiversx/mx-sdk-dapp/pull/878)
```
Comment (arhtudormorar) [inline CHANGELOG.md:10]: Added support for usernames

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx [hunk 4]
```diff
@@ -1,25 +1,80 @@
...
+  receiver,
+  scamReport
+}: ConfirmReceiverPropsType) => {
+  const isSmartContract = isContract(receiver);
+
+  const {
+    account: usernameAccount,
+    loading: usernameAccountLoading,
+    error: usernameAccountError
+  } = useGetAccountFromApi(isSmartContract ? null : receiver);
```
Comment (arhtudormorar) [inline src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx:34]: check out null

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx [hunk 5]
```diff
@@ -1,25 +1,80 @@
...
+  );
 
   return (
     <div className={styles.receiver}>
       <span className={styles.label}>Receiver</span>
 
-      {receiver && (
-        <span className={styles.value} data-testid='confirmReceiver'>
-          {receiver}
+      {receiver && !usernameAccountLoading ? (
```
Comment (arhtudormorar) [inline src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx:45]: usernameAccountLoading? first:second

check out ternary in ternary (maybe split above in smaller components with Memo)

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/components/ReceiverSubValue/ReceiverSubValue.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  if (hasUsername && !isSmartContract) {
+    return (
+      <span className={styles.subValue}>
+        <Trim text={receiver} className={styles.subValueTrim} />
+        <CopyButton text={receiver} className={styles.subValueCopy} />
+      </span>
+    );
+  }
+
+  if (isSmartContract && !usernameAccountLoading) {
```
Comment (arhtudormorar) [inline src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/components/ReceiverSubValue/ReceiverSubValue.tsx:33]: if usernameAccountLoading return null
if isSmartcontract return ...
...

File: src/hooks/account/useGetAccountFromApi.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import { useEffect, useState } from 'react';
+
+import { getAccountFromApi } from 'apiCalls';
+import { AccountType } from 'types';
+
+export const useGetAccountFromApi = (address: string | null) => {
```
Comment (arhtudormorar) [inline src/hooks/account/useGetAccountFromApi.ts:6]: use swr for this. It makes it easyer with null

File: CHANGELOG.md (hunk 3 above)
Comment (MiroMargineanu) [reply CHANGELOG.md]: Added. 👍

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx (hunk 5 above)
Comment (MiroMargineanu) [reply src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx]: Updated. 👍

**Review by razvantomegea — APPROVED** [review]

File: src/hooks/account/useGetAccountFromApi.ts (hunk 7 above)
Comment (MiroMargineanu) [reply src/hooks/account/useGetAccountFromApi.ts]: Added. 👍

File: src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx (hunk 4 above)
Comment (MiroMargineanu) [reply src/UI/SignTransactionsModals/SignWithDeviceModal/components/components/ConfirmReceiver/ConfirmReceiver.tsx]: Removed. 👍

**Pull request merged**